File: src/episode-update-publish.py
import csv
import os
import requests
from datetime import datetime

# Constants
csv_file_path = './config/scotus.csv'
TRANSISTOR_API_URL = "https://api.transistor.fm/v1/episodes"
TRANSISTOR_API_KEY = os.getenv('TRANSISTOR_API_KEY')
headers = {"x-api-key": TRANSISTOR_API_KEY, "Content-Type": "application/json"}

def publish_episode(episode_id):
    """
    Publish an episode on Transistor.fm
    """
    url = f"{TRANSISTOR_API_URL}/{episode_id}/publish"

    response = requests.post(url, headers=headers)
    return response.json()

def process_csv(csv_path):
    """
    Process the CSV file and publish episodes if the argument date has passed and they are not published.
    """
    with open(csv_path, mode='r', encoding='utf-8') as file:
        reader = csv.DictReader(file)

        for row in reader:
            argument_date = datetime.strptime(row["Argument Date"], "%Y-%m-%d").date()
            current_date = datetime.now().date()

            if 'Episode ID' in row and row['Episode ID'] and argument_date < current_date:
                # Assuming there's a column 'Published' to check if the episode is already published
                if row.get('Published', 'no').lower() != 'yes':
                    response = publish_episode(row['Episode ID'])
                    print(f"Episode {row['Episode ID']} published: {response}")

# Main execution
process_csv(csv_file_path)

File: src/go/episode-update-publish.go
package main

import (
    "bufio"
    "encoding/csv"
    "encoding/json"
    "fmt"
    "net/http"
    "os"
    "time"
)

type Config struct {
    CSVFilePath       string
    TransistorAPIURL  string
    TransistorAPIKey  string
}

func LoadConfig() Config {
    return Config{
        CSVFilePath:       "./config/scotus.csv",
        TransistorAPIURL:  "https://api.transistor.fm/v1/episodes",
        TransistorAPIKey:  os.Getenv("TRANSISTOR_API_KEY"),
    }
}

func publishEpisode(episodeID, apiURL, apiKey string) (json.RawMessage, error) {
    url := fmt.Sprintf("%s/%s/publish", apiURL, episodeID)

    client := &http.Client{}
    req, err := http.NewRequest("POST", url, nil)
    if err != nil {
        return nil, err
    }
    req.Header.Set("x-api-key", apiKey)
    req.Header.Set("Content-Type", "application/json")

    resp, err := client.Do(req)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    var response json.RawMessage
    err = json.NewDecoder(resp.Body).Decode(&response)
    return response, err
}

func processCSV(csvPath, apiURL, apiKey string) error {
    file, err := os.Open(csvPath)
    if err != nil {
        return err
    }
    defer file.Close()

    reader := csv.NewReader(bufio.NewReader(file))
    rows, err := reader.ReadAll()
    if err != nil {
        return err
    }

    for _, row := range rows[1:] {
        argumentDate, err := time.Parse("2006-01-02", row[1]) // Assuming 'Argument Date' is the second column
        if err != nil {
            fmt.Println("Error parsing date:", err)
            continue
        }

        if row[2] != "" && argumentDate.Before(time.Now()) && row[3] != "yes" { // Assuming 'Published' is the fourth column
            response, err := publishEpisode(row[2], apiURL, apiKey) // Assuming 'Episode ID' is the third column
            if err != nil {
                fmt.Println("Error publishing episode:", err)
                continue
            }
            fmt.Printf("Episode %s published: %s\n", row[2], string(response))
        }
    }
    return nil
}

func main() {
    config := LoadConfig()

    if err := processCSV(config.CSVFilePath, config.TransistorAPIURL, config.TransistorAPIKey); err != nil {
        fmt.Println("Error processing CSV file:", err)
    }
}
